component/status/status.go

```diff
@@ -0,0 +1,131 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package status
+
+import (
+	"errors"
+	"time"
+
+	"go.opentelemetry.io/collector/component/id"
+)
+
+// ComponentEventType represents an enumeration of component status event types.
+type ComponentEventType int
+
+const (
+	// ComponentOK indicates the component is functioning normally.
+	ComponentOK ComponentEventType = iota
+	// ComponentError indicates the component is in erroneous state.
+	ComponentError
+)
+
+// Source component that reports a status about itself.
+// The implementation of this interface must be comparable to be useful as a map key.
+type Source interface {
+	ID() id.ID
+}
+
+// ComponentEvent is a status event produced by a component to communicate its status to
+// registered listeners.
+// An event can signal that a component is working normally (status.ComponentOK),
+// or that it is in an error state (status.ComponentError). An error status may
+// optionally include an error object to provide additional insight to
+// registered listeners.
+type ComponentEvent struct {
+	eventType ComponentEventType
+	timestamp time.Time
+	source    Source
+	err       error
+}
+
+// Type returns the event type.
+func (ev *ComponentEvent) Type() ComponentEventType {
+	return ev.eventType
+}
+
+// Source is the component that is reporting its status.
+func (ev *ComponentEvent) Source() Source {
+	return ev.source
+}
+
+// Timestamp returns the time of the event.
+func (ev *ComponentEvent) Timestamp() time.Time {
+	return ev.timestamp
+}
+
+// Err returns the error associated with the ComponentEvent.
+func (ev *ComponentEvent) Err() error {
+	return ev.err
+}
+
+// ComponentEventOption applies options to a ComponentEvent.
+type ComponentEventOption func(*ComponentEvent) error
```

Comment:
either make the arg of the func private, or use an interface with a private func.